mtags/src/main/scala/scala/meta/internal/pc/TypeDefinitionProvider.scala

```diff
@@ -0,0 +1,152 @@
+package scala.meta.internal.pc
+
+import org.eclipse.{lsp4j => l}
+import scala.meta.internal.mtags.MtagsEnrichments._
+import scala.meta.pc.OffsetParams
+import scala.meta.internal.jdk.CollectionConverters._
+import scala.meta.internal.metals.CompilerOffsetParams
+
+class TypeDefinitionProvider(val compiler: MetalsGlobal, params: OffsetParams) {
+  import compiler._
+
+  def typedTree: Option[(RichCompilationUnit, Position, Tree)] = {
+    if (params.isWhitespace) {
+      None
+    } else {
+      val unit = addCompilationUnit(
+        code = params.text(),
+        filename = params.filename(),
+        cursor = None
+      )
+      val pos = unit.position(params.offset())
+      val tree = typedHoverTreeAt(pos)
+      Some(unit, pos, tree)
+    }
+  }
+
+  def typeSymbol: Option[(RichCompilationUnit, Position, Symbol)] = {
+    val tri = typedTree
+    tri match {
+      case Some((t1, t2, tree))
+          if tree.symbol.isTypeSymbol || tree.symbol.isMethod || tree.symbol.isConstructor =>
+        Some(t1, t2, tree.symbol)
+      case Some((t1, t2, tree)) if tree.tpe.isDefined =>
+        Some(t1, t2, tree.tpe.typeSymbol)
+      case Some((t1, t2, tree)) if tree.children.nonEmpty =>
+        Some(t1, t2, tree.children.head.tpe.typeSymbol)
+      case Some((t1, t2, tree)) =>
+        val expTree = expandRangeToEnclosingApply(tree.pos)
+        if (expTree.tpe.isDefined) Some(t1, t2, expTree.tpe.typeSymbol)
+        else None
+      case _ => None
+    }
+  }
+
+  def typeDefinition: List[l.Location] = {
+    typeSymbol match {
+      case Some((_, _, sym))
```

Comment:
Would just flatMap over typeDefinition result.

Reply:
What do you mean by that?